### PR TITLE
Enable Lisa on AWS to execute the network and storge performance tests

### PR DESCRIPTION
**Pull request by rainierside**

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -38,6 +38,8 @@ pycdlib = "^1.12.0"
 randmac = "^0.1"
 toml = "^0.10.2"
 Pillow = {version="^9.0.0", optional=true}
+boto3 = "1.21.37"
```
Comment (squirrelsc) [inline pyproject.toml]: Mark aws related package with optional and create an "aws" extra for it

File: lisa/sut_orchestrator/aws/common.py [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+    username: str = ""
+    password: str = ""
+    private_key_file: str = ""
+
+
+@dataclass_json()
+@dataclass
+class AwsVmMarketplaceSchema:
+    # amazon/AWS Deep Learning AMI GPU CUDA 11.2.1 (Ubuntu 20.04) 20220208
+    imageid: str = "ami-097324d9d7113bccb"
```
Comment (squirrelsc) [inline lisa/sut_orchestrator/aws/common.py]: Use a cheaper one as default image

File: lisa/sut_orchestrator/aws/common.py [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+    )
+    data_disk_iops: int = 500
+    data_disk_size: int = 32
+    disk_type: str = ""
+
+    # for marketplace image, which need to accept terms
+    _marketplace: InitVar[Optional[AwsVmMarketplaceSchema]] = None
+
+    @property
+    def marketplace(self) -> Optional[AwsVmMarketplaceSchema]:
```
Comment (squirrelsc) [inline lisa/sut_orchestrator/aws/common.py]: What's the format of marketplace image id? If it's not complex as Azure, it doesn't need this logic.

File: lisa/sut_orchestrator/aws/features.py [hunk 4]
```diff
@@ -0,0 +1,503 @@
...
+        platform: AwsPlatform = self._platform  # type: ignore
+        ec2_client = platform._ec2_client
+        ec2_client.start_instances(InstanceIds=[self._intsance_id])
+        ec2_client.get_waiter("instance_running").wait()
+
+    def _restart(self, wait: bool = True) -> None:
+        platform: AwsPlatform = self._platform  # type: ignore
+        ec2_client = platform._ec2_client
+        ec2_client.reboot_instances(InstanceIds=[self._intsance_id])
+        ec2_client.get_waiter("instance_running").wait()
```
Comment (squirrelsc) [inline lisa/sut_orchestrator/aws/features.py]: Use the `wait` argument to decide whether waiting in this method.

File: lisa/sut_orchestrator/aws/features.py [hunk 5]
```diff
@@ -0,0 +1,503 @@
...
+
+
+_disk_type_mapping: Dict[schema.DiskType, str] = {
+    schema.DiskType.PremiumSSDLRS: "io1",
+    schema.DiskType.StandardHDDLRS: "st1",
+    schema.DiskType.StandardSSDLRS: "gp3",
+}
+
+# disk types are ordered by commonly and cost. The earlier is lower cost.
+_ordered_disk_types: List[schema.DiskType] = [
```
Comment (squirrelsc) [inline lisa/sut_orchestrator/aws/features.py]: If some items are shared between Azure and Aws. We can create a common.py under sut_orchestrator to put them.

File: pyproject.toml (hunk 1 above)
Comment (rainierside) [reply pyproject.toml]: Updated, thanks

File: lisa/sut_orchestrator/aws/common.py (hunk 2 above)
Comment (rainierside) [reply lisa/sut_orchestrator/aws/common.py]: Updated, thanks

File: lisa/sut_orchestrator/aws/common.py (hunk 3 above)
Comment (rainierside) [reply lisa/sut_orchestrator/aws/common.py]: Updated, thanks

File: lisa/sut_orchestrator/aws/features.py (hunk 4 above)
Comment (rainierside) [reply lisa/sut_orchestrator/aws/features.py]: Updated, thanks

File: lisa/sut_orchestrator/aws/features.py (hunk 5 above)
Comment (rainierside) [reply lisa/sut_orchestrator/aws/features.py]: When rebase to latest change on main branch, these definition is moved to the common schema.py.

File: lisa/sut_orchestrator/aws/features.py (hunk 4 above)
Comment (squirrelsc) [reply lisa/sut_orchestrator/aws/features.py]: The _start/_stop should use the `wait` argument as well.

File: lisa/sut_orchestrator/aws/common.py [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+def wait_operation(operation: Any, time_out: int = sys.maxsize) -> Any:
+    timer = create_timer()
+    while time_out > timer.elapsed(False):
+        check_cancelled()
+        if operation.done():
+            break
+        operation.wait(1)
+    if time_out < timer.elapsed():
+        raise Exception(
+            f"timeout on wait Azure operation completed after {time_out} seconds."
```
Comment (squirrelsc) [inline lisa/sut_orchestrator/aws/common.py]: It's AWS, not Azure.

File: lisa/sut_orchestrator/aws/common.py [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+
+def get_node_context(node: Node) -> NodeContext:
+    return node.get_context(NodeContext)
+
+
+def get_environment_context(environment: Environment) -> EnvironmentContext:
+    return environment.get_context(EnvironmentContext)
+
+
+def wait_operation(operation: Any, time_out: int = sys.maxsize) -> Any:
```
Comment (squirrelsc) [inline lisa/sut_orchestrator/aws/common.py]: This method is used to simplify the wait operation with timeout support. It should be used as many as possible.

File: lisa/sut_orchestrator/aws/features.py (hunk 4 above)
Comment (rainierside) [reply lisa/sut_orchestrator/aws/features.py]: Fixed.

File: lisa/sut_orchestrator/aws/common.py (hunk 6 above)
Comment (rainierside) [reply lisa/sut_orchestrator/aws/common.py]: Updated

File: lisa/sut_orchestrator/aws/common.py (hunk 7 above)
Comment (rainierside) [reply lisa/sut_orchestrator/aws/common.py]: Remove this wait_operation function as it is not used. There is some wait method in AWS apis.

File: lisa/sut_orchestrator/aws/platform_.py [hunk 8]
```diff
@@ -158,11 +159,11 @@ def _initialize(self, *args: Any, **kwargs: Any) -> None:
 
         self._aws_runbook = aws_runbook
         self._ec2_client = boto3.client("ec2")
```
Comment (squirrelsc) [inline lisa/sut_orchestrator/aws/platform_.py:197]: Remove this line? So every new instance will be created at needed.

`self._ec2_client = boto3.client("ec2")`

File: lisa/sut_orchestrator/aws/platform_.py [hunk 9]
```diff
@@ -125,6 +137,27 @@ class AwsPlatformSchema:
...
+                "aws_default_region",
+                "security_group_name",
+                "key_pair_name",
+                "locations",
+                "log_level",
+            ],
+        )
+
+        if self.aws_access_key_id:
+            add_secret(self.aws_access_key_id, mask=PATTERN_GUID)
```
Comment (squirrelsc) [inline lisa/sut_orchestrator/aws/platform_.py]: If it's not GUID format, it doesn't need to specify the mask as GUID. It can be left as default.

**Review by squirrelsc — APPROVED** [review]

**Pull request merged**